### DISCUSSION TITLE
How to call the mean-squared displacement of atoms within NCrystal

### DISCUSSION THREAD

**Discussion by XuShuqi7** [mctools/ncrystal#83] [Q&A] [answered]
Hi Everyone,

my name is Shuqi Xu, I am currently participating in the HighNESS project and implementing the neutron magnetic scattering model in a plugin in NCrystal.

I would like to call the mean-squared displacement (MSD) calculated by NCrystal in my function. After checking the source files of NCrystal, it seems that it can be called by:

  double msd = 0.;
  if ( info.hasAtomMSD() ) msd = info.getAtomInfos().front().msd();

But I got the error message:

-- Configuring done
-- Generating done
-- Build files have been written to: /Users/xushuqi/Downloads/My_C++/ncrystal-3.0.0
Consolidate compiler generated dependencies of target NCrystal
[  1%] Building CXX object CMakeFiles/NCrystal.dir/ncplugin-MagScat_V2/src/NCParamagneticScatter.cc.o
/Users/xushuqi/Downloads/My_C++/ncrystal-3.0.0/ncplugin-MagScat_V2/src/NCParamagneticScatter.cc:105:62: error: assigning to 'double' from incompatible type 'const Optional<double>'
  if ( info.hasAtomMSD() ) msd = info.getAtomInfos().front().msd();
                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
1 error generated.
make[2]: *** [CMakeFiles/NCrystal.dir/ncplugin-MagScat_V2/src/NCParamagneticScatter.cc.o] Error 1
make[1]: *** [CMakeFiles/NCrystal.dir/all] Error 2
make: *** [all] Error 2

It seems that the issue is related to the data type but I cannot solve it. I would like to ask for your help if anyone knows how to fix this, thx 🙏.

Best regards,
Shuqi

To be clearer, the whole function is put here:

NCP::ParamagneticScatter NCP::ParamagneticScatter::createFromInfo( const NC::Info& info )
{
  //Parse the content of our custom section. In case of syntax errors, we should
  //raise BadInput exceptions, to make sure users gets understandable error
  //messages. We should try to avoid other types of exceptions.

  //Get the relevant custom section data (and verify that there are not multiple
  //such sections in the input data):
  if ( info.countCustomSections( pluginNameUpperCase() ) != 1 )
    NCRYSTAL_THROW2(BadInput,"Multiple @CUSTOM_"<<pluginNameUpperCase()<<" sections are not allowed");
  auto data = info.getCustomSection( pluginNameUpperCase() );

  // data is here a vector of lines, and each line is a vector of words. In our
  // case, we want to accept sections of the form (units are barn, angstrom^-1,
  // eV, boolean, as is usual in NCrystal):
  //
  // @CUSTOM_<ourpluginname>
  //    <sigmavalue> <hwhm value> <D_const value> <down-scattering option>
  //

  //Verify we have exactly one line and four words:
  if ( data.size() != 1 || data.at(0).size()!=4 )
    NCRYSTAL_THROW2(BadInput,"Data in the @CUSTOM_"<<pluginNameUpperCase()
                    <<" section should be four numbers on a single line");

  //Parse and validate values:
  double sigma, hwhm, D_const;
  if ( ! NC::safe_str2dbl( data.at(0).at(0), sigma )
       || ! NC::safe_str2dbl( data.at(0).at(1), hwhm )
       || ! NC::safe_str2dbl( data.at(0).at(2), D_const )
       || ! (sigma>0.0) || ! (hwhm>0.0) || ! (D_const>0.0) )
    NCRYSTAL_THROW2( BadInput,"Invalid values specified in the @CUSTOM_"<<pluginNameUpperCase()
                     <<" sigma, hwhm, D_const (should be three positive floating point values)" );
    
  int mag_scat;
  if ( ! NC::safe_str2int( data.at(0).at(3), mag_scat )
       || ! (mag_scat > -2) || ! (mag_scat < 2) )
    NCRYSTAL_THROW2( BadInput,"Invalid value specified in the @CUSTOM_"<<pluginNameUpperCase()
                     <<" mag_scat (must be -1 (for down-scattering) or 1 (for up-scattering) or 0 (for elastic scattering)" );
  
  //Getting the temperature
  double temperature = info.getTemperature().get();
    
  //Getting the mean-squared displacement (MSD)
  double msd = 0.;
  if ( info.hasAtomMSD() ) msd = info.getAtomInfos().front().msd();

  //Parsing done! Create and return our model:
  return ParamagneticScatter(sigma,hwhm,D_const,temperature,mag_scat);
}

**Comment by tkittel** [comment, marked as answer]
Hi @XuShuqi7 ,

As you can see on [NCInfoTypes.hh](https://github.com/mctools/ncrystal/blob/master/ncrystal_core/include/NCrystal/NCInfoTypes.hh) the `msd()` method returns an `Optional<double>` rather than a raw `double`. The Optional class is our C++11 compatible version of [std::optional](https://en.cppreference.com/w/cpp/utility/optional) which is defined in [NCDefs.hh](https://github.com/mctools/ncrystal/blob/master/ncrystal_core/include/NCrystal/NCDefs.hh). This is a simple container class which may or may not contain an actual value. One can use the `.has_value()` method to check if there is a value (although in your case the check for `hasAtomMSD()` already did the `.has_value()` check internally. If there is a value, the `.value()` method can be used to access it.

Long story short, instead of:

```c++
double msd = 0;
if ( info.hasAtomMSD() )
  msd = info.getAtomInfos().front().msd(); //Wrong, won't compile!
```

you can write:

```c++
double msd = 0;
if ( info.hasAtomMSD() )
  msd = info.getAtomInfos().front().msd().value();
```

Of course, it looks like your code is assuming mono-atomic materials, but that is another issue :-)

Cheers,
Thomas

ps. If you add three backticks around your code blocks in github messages, it becomes much more readable :-). See more on https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by XuShuqi7** [reply]
Hi Thomas,

thank you for your prompt reply. According to your advice, I succeeded in compiling my plugin!

Yes it is assumed monatomic in my case. I will see in details for a more general application. 

The three backticks are noted and I will use them for the next time😊.

Best regards,
Shuqi
